Tools/scripts/update_file.py

```diff
@@ -6,23 +6,51 @@
 actually change the in-tree generated code.
 """
 
+import contextlib
 import os
+import os.path
 import sys
 
 
-def main(old_path, new_path):
-    with open(old_path, 'rb') as f:
+@contextlib.contextmanager
+def updating_file_with_tmpfile(filename, tmpfile=None):
+    """A context manager for updating a file via a temp file.
+
+    The context manager provides two open files: the source file open
+    for reading, and the temp file, open for writing.
+
+    Upon exiting: both files are closed, and the source file is replaced
+    with the temp file.
+    """
+    # XXX Optionally use tempfile.TemporaryFile?
+    if not tmpfile:
+        tmpfile = filename + '.tmp'
+    elif os.path.isdir(tmpfile):
+        tmpfile = os.path.join(tmpfile, filename + '.tmp')
+
+    try:
+        with open(tmpfile, 'w') as outfile:
+            with open(filename) as infile:
+                yield infile, outfile
+        os.replace(tmpfile, filename)
+    finally:
+        if os.path.exists(tmpfile):
+            os.remove(tmpfile)
+
+
+def update_file_with_tmpfile(filename, tmpfile):
```

Comment:
Why not define this in terms of the context manager?

Reply:
I've gone the other direction, implementing the context manager in terms of `update_file_with_tmpfile()`.